torch/_inductor/cudagraph_trees.py

```diff
@@ -0,0 +1,906 @@
+    return manager.add_function(model, inputs, static_input_idxs)
+
+
+def is_live(weak_ref):
+    if weak_ref is None:
+        return False
+    return weak_ref() is not None
+
+
+class TensorWeakRef:
+    """
+    Wrapper around a weak ref of a Tensor that handles the _fix_weakref() call required
+    when unwrapping a Tensor weakref.
+    """
+
+    ref: WeakRef[Tensor]
+
+    def __init__(self, tensor: Tensor):
+        assert isinstance(tensor, Tensor)
+        self.ref = weakref.ref(tensor)
+
+    def __call__(self):
+        out = self.ref()
+        if out is None:
+            return out
+        assert isinstance(out, Tensor)
+        out._fix_weakref()
+        return out
+
+
+def is_cuda_tensor(x):
+    return isinstance(x, torch.Tensor) and x.device.type == "cuda"
+
+
+class CUDAGraphNode:
+    """
+    A single recording of a function into a CUDA Graph. Recordings of CUDA Graphs share a single memory pool
+    and are structured into a tree, where there is a single recording that can precede it (parent) and multiple
+    subsequent recordings that may follow (children). A node will have no parent if it is the first recording
+    in a tree; i.e., when it is first recorded, there are no live tensors from a previous recording which
+    would force a dependency.
+
+    On first recording, all of the live tensors in the current CUDA Graph Node path will be
+    reflected in the corresponding private pool. On subsequent executions, the caching allocator
+    is unaffected when the graph is replayed.
+
+    In order to support recording a subsequent cuda graph recording after execution of this graph,
+    we checkpoint the state of the memory pool so that it may later be resumed.
+
+    See [setCheckpointPoolState] for further explanation, as well as
+    https://user-images.githubusercontent.com/13564/222815509-374f3400-f83d-4f7d-8fa6-4a092b3250bb.png
+    """
+
+    def __init__(
+        self,
+        wrapped_function: WrappedFunction,
+        id: GraphID,
+        parent: Optional[CUDAGraphNode],
+        inputs: List[Tensor],
+        cuda_graphs_pool: Tuple[int, int],
+    ):
+        assert isinstance(inputs, (list, tuple))
+
+        self.wrapped_function = wrapped_function
+        self.id = id
+
+        # if this is a root parent will be None. use weakref to prevent reference cycle
+        self._parent = weakref.ref(parent) if parent is not None else None
+        self.cuda_graphs_pool = cuda_graphs_pool
+
+        # A single wrapped function may be recorded multiple times if memory patterns or
+        # invariants change from one execution to the next
+        self.children: Dict[FunctionID, List[CUDAGraphNode]] = defaultdict(list)
+
+        self.device: int = next(
+            (x.device.index for x in inputs if is_cuda_tensor), None
+        )
+
+        # we preserve a single reference to executed outputs that is then referenced
+        # in children to avoid children having to chase parent pointers in the hot path
```

Comment:
There seems to be some invariant here, it would be nice if it was explicitly written out

Reply:
In particular, according to the comment, it should be possible to recompute the output weakrefs by chasing parent pointers. We should implement this and have a debug mode to actually check that this is consistent.

Reply:
Can implement this. I dont think there is any invariant here besides what is commented.